datalad/core/local/diff.py

```diff
@@ -0,0 +1,335 @@
+                    logger=lgr)
+                continue
+            if path_under_rev_dataset(ds, str_path) is None:
+                # nothing we support handling any further
+                # there is only a single refds
+                yield dict(
+                    path=str_path,
+                    refds=ds.path,
+                    action='diff',
+                    status='error',
+                    message=(
+                        "dataset containing given paths is not underneath "
+                        "the reference dataset %s: %s",
+                        ds, str_path),
+                    logger=lgr,
+                )
+                continue
+
+            ps.append(p)
+        path = ps
+
+    # TODO we might want to move away from the single-pass+immediate-yield
+    # paradigm for this command. If we gather all information first, we
+    # could do post-processing and detect when a file (same gitsha, or same
+    # key) was copied/moved from another dataset. Another command (e.g.
+    # rev-save) could act on this information and also move/copy
+    # availability information or at least enhance the respective commit
+    # message with cross-dataset provenance info
+
+    # cache to help avoid duplicate status queries
+    content_info_cache = {}
+    for res in _diff_ds(
+            ds,
+            fr,
+            to,
+            constant_refs,
+            recursion_limit
+            if recursion_limit is not None and recursive
+            else -1 if recursive else 0,
+            # TODO recode paths to repo path reference
+            origpaths=None if not path else OrderedDict(path),
+            untracked=untracked,
+            annexinfo=annex,
+            eval_file_type=True,
+            cache=content_info_cache):
+        res.update(
+            refds=ds.path,
+            logger=lgr,
+            action='diff',
+        )
+        yield res
+
+
+def _diff_ds(ds, fr, to, constant_refs, recursion_level, origpaths, untracked,
+             annexinfo, eval_file_type, cache):
+    if not ds.repo:
+        # asked to query a subdataset that is not available
+        lgr.debug("Skip diff of unavailable subdataset: %s", ds)
+        return
+
+    repo_path = ds.repo.pathobj
+    # filter and normalize paths that match this dataset before passing them
+    # onto the low-level query method
+    paths = None if origpaths is None \
+        else OrderedDict(
+            (repo_path / p.relative_to(ds.pathobj), goinside)
+            for p, goinside in iteritems(origpaths)
+            if ds.pathobj in p.parents or (p == ds.pathobj and goinside)
+        )
+    try:
+        lgr.debug("diff %s from '%s' to '%s'", ds, fr, to)
+        diff_state = ds.repo.diffstatus(
+            fr,
+            to,
+            paths=None if not paths else [p for p in paths],
+            untracked=untracked,
+            eval_file_type=eval_file_type,
+            _cache=cache)
+    except ValueError as e:
```

Comment:
I would say we better define a custom `InvalidReferenceError(ValueError)` (to be thrown in `GitRepo.get_content_info`) and avoid matching string representation of an exception here - fragile etc.

Reply:
If you know how to do it without string matching, please let me know. Git doesn't give me an exception.

Reply:
https://github.com/datalad/datalad/issues/3375